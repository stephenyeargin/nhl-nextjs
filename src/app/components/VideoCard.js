'use client';

import React, { useEffect, useState } from 'react';
import Image from 'next/image';
import Link from 'next/link';
import { PropTypes } from 'prop-types';
import { formatLocalizedDate, formatLocalizedTime, formatMarkdownContent } from '../utils/formatters';

const MissingThumbnail = ({ size }) => (
  <div className={`w-full bg-gray-200 flex items-center justify-center aspect-square mb-2 ${size === 'medium' ? 'aspect-square' : 'aspect-video'}`}>
    <Image
      src="https://assets.nhle.com/logos/nhl/svg/NHL_light.svg"
      width="180"
      height="180"
      alt="Story Photo"
      className=""
    />
  </div>
);

MissingThumbnail.propTypes = {
  size: PropTypes.string,
};

const VideoCard = ({ item, size, className, handleCardClick }) => {
  const [blurDataURL, setBlurDataURL] = useState('data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==');

  useEffect(() => {
    const fetchBlurDataURL = async () => {
      const data = await fetch(item.thumbnail?.templateUrl.replace('{formatInstructions}', 't_ratio16_9-size10/f_png'));
      const base64 = await data.blob().then((blob) => new Promise((resolve) => {
        const reader = new FileReader();
        reader.readAsDataURL(blob);
        reader.onloadend = () => resolve(reader.result);
      }));
      setBlurDataURL(base64);
    };
    fetchBlurDataURL();
  }, [item.thumbnail]);

  if (blurDataURL === null) {
    return <></>;
  }

  if (size === 'small') {
    return (
      <div className={className}>
        <Link onClick={handleCardClick} href={`/video/${item.slug}`} className="grid grid-cols-3 gap-4 items-center">
          <div className="col-span-1">
            {item.thumbnail ? (
              <Image
                src={item.thumbnail?.thumbnailUrl}
                width="416"
                height="416"
                alt="Story Photo"
                className="mb-2 w-full"
                loading="lazy"
              />
            ) : (
              <MissingThumbnail />
            )}
          </div>
          <div className="col-span-2">
            <h2 className="text-base font-bold">{item.headline || item.title}</h2>
          </div>
        </Link>
      </div>
    );
  }

  if (size === 'medium') {
    return(
      <div className={className}>
        <p className="hidden" suppressHydrationWarning>{formatLocalizedDate(item.contentDate)} {formatLocalizedTime(item.contentDate)}</p>
        <Link onClick={handleCardClick} href={`/video/${item.slug}`} className="">
          {item.thumbnail ? (
            <Image
              src={item.thumbnail?.thumbnailUrl}
              width="832"
              height="468"
              alt="Story Photo"
              className="mb-2 w-full"
              loading="lazy"
            />
          ) : (
            <MissingThumbnail size="medium" />
          )}
        </Link>
        <Link onClick={handleCardClick} href={`/video/${item.slug}`} className="">
          <h2 className="text-xl font-bold">{item.headline || item.title}</h2>
        </Link>
        <div className="text-justify line-clamp-3 text-sm" dangerouslySetInnerHTML={{ __html: formatMarkdownContent(item.summary) }} />
        <Link onClick={handleCardClick} href={`/video/${item.slug}`} className="block font-bold py-3 underline">Watch</Link>
      </div>
    );
  }

  if (size === 'large') {
    return (
      <div className={`md:relative ${className}`}>
        <Link onClick={handleCardClick} href={`/video/${item.slug}`} className="">
          <Image
            src={item.thumbnail?.templateUrl.replace('{formatInstructions}', 't_ratio16_9-size40/f_png')}
            width="832"
            height="468"
            alt="Story Photo"
            className="w-full"
            placeholder="blur"
            blurDataURL={blurDataURL}
            loading="lazy"
          />
        </Link>
        <div className="md:absolute md:bottom-0 md:right-0 md:left-0 md:p-5 md:bg-black/70 md:text-white">
          <Link onClick={handleCardClick} href={`/video/${item.slug}`} className="">
            <h2 className="text-2xl font-bold">{item.headline || item.title}</h2>
          </Link>
          <div className="text-justify line-clamp-3 text-sm" dangerouslySetInnerHTML={{ __html: formatMarkdownContent(item.summary)}} />
          <Link href={`/video/${item.slug}`} className="block font-bold py-3 underline">Read Story</Link>
        </div>
      </div>
    );
  }

  return(
    <div className={className}>
      <p className="hidden" suppressHydrationWarning>{formatLocalizedDate(item.contentDate)} {formatLocalizedTime(item.contentDate)}</p>
      <Link onClick={handleCardClick} href={`/video/${item.slug}`} className="">
        {item.thumbnail ? (
          <Image
            src={item.thumbnail?.templateUrl.replace('{formatInstructions}', 't_ratio16_9-size20/f_png')}
            width="832"
            height="468"
            alt="Story Photo"
            className="mb-2 w-full"
            placeholder="blur"
            blurDataURL={blurDataURL}
            loading="lazy"
          />
        ) : (
          <MissingThumbnail />
        )}
        <h2 className="text-xl font-bold">{item.headline || item.title}</h2>
        <div className="text-sm mt-3">{item.fields.description}</div>
      </Link>
    </div>
  );
};

VideoCard.propTypes = {
  item: PropTypes.object.isRequired,
  className: PropTypes.string,
  size: PropTypes.string,
  handleCardClick: PropTypes.func,
};

export default VideoCard;
