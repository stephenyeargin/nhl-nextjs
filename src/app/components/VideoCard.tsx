'use client';

import React, { useEffect, useState } from 'react';
import Image from 'next/image';
import Link from 'next/link';
import {
  formatLocalizedDate,
  formatLocalizedTime,
  formatMarkdownContent,
} from '../utils/formatters';
import type { VideoItemBase } from '@/app/types/video';

interface MissingThumbnailProps {
  size?: 'small' | 'medium' | 'large';
}

interface VideoCardProps {
  item: VideoItemBase;
  className?: string;
  size?: 'small' | 'medium' | 'large';
  handleCardClick?: (_e: React.MouseEvent<HTMLAnchorElement | HTMLDivElement>) => void; // consumer may use event
}

const MissingThumbnail = ({ size }: MissingThumbnailProps) => (
  <div
    className={`w-full bg-gray-200 flex items-center justify-center aspect-square mb-2 ${size === 'medium' ? 'aspect-square' : 'aspect-video'}`}
  >
    <Image
      src="https://assets.nhle.com/logos/nhl/svg/NHL_light.svg"
      width="180"
      height="180"
      alt="Story Photo"
      className=""
    />
  </div>
);

const VideoCard = ({ item, size, className, handleCardClick }: VideoCardProps) => {
  const [blurDataURL, setBlurDataURL] = useState(
    'data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=='
  );

  useEffect(() => {
    const fetchBlurDataURL = async () => {
      if (!item.thumbnail?.templateUrl) {
        return;
      }
      const data = await fetch(
        item.thumbnail.templateUrl.replace('{formatInstructions}', 't_ratio16_9-size10/f_png')
      );
      const base64 = await data.blob().then(
        (blob) =>
          new Promise<string>((resolve) => {
            const reader = new FileReader();
            reader.readAsDataURL(blob);
            reader.onloadend = () => resolve(reader.result as string);
          })
      );
      setBlurDataURL(base64);
    };
    fetchBlurDataURL();
  }, [item.thumbnail]);

  if (blurDataURL === null) {
    return <></>;
  }

  if (size === 'small') {
    return (
      <div className={className}>
        <Link
          onClick={handleCardClick}
          href={`/video/${item.slug}`}
          className="grid grid-cols-3 gap-4 items-center"
        >
          <div className="col-span-1">
            {item.thumbnail ? (
              <Image
                src={
                  item.thumbnail?.thumbnailUrl ||
                  'https://assets.nhle.com/logos/nhl/svg/NHL_light.svg'
                }
                width="416"
                height="416"
                alt="Story Photo"
                className="mb-2 w-full"
                loading="lazy"
              />
            ) : (
              <MissingThumbnail />
            )}
          </div>
          <div className="col-span-2">
            <h2 className="text-base font-bold">{item.headline || item.title}</h2>
          </div>
        </Link>
      </div>
    );
  }

  if (size === 'medium') {
    return (
      <div className={className}>
        <p className="hidden" suppressHydrationWarning>
          {formatLocalizedDate(item.contentDate)} {formatLocalizedTime(item.contentDate)}
        </p>
        <Link onClick={handleCardClick} href={`/video/${item.slug}`} className="">
          {item.thumbnail ? (
            <Image
              src={
                item.thumbnail?.thumbnailUrl ||
                'https://assets.nhle.com/logos/nhl/svg/NHL_light.svg'
              }
              width="832"
              height="468"
              alt="Story Photo"
              className="mb-2 w-full"
              loading="lazy"
            />
          ) : (
            <MissingThumbnail size="medium" />
          )}
        </Link>
        <Link onClick={handleCardClick} href={`/video/${item.slug}`} className="">
          <h2 className="text-xl font-bold">{item.headline || item.title}</h2>
        </Link>
        <div
          className="text-justify line-clamp-3 text-sm"
          dangerouslySetInnerHTML={{ __html: formatMarkdownContent(item.summary) }}
        />
        <Link
          onClick={handleCardClick}
          href={`/video/${item.slug}`}
          className="block font-bold py-3 underline"
        >
          Watch
        </Link>
      </div>
    );
  }

  if (size === 'large') {
    return (
      <div className={`md:relative ${className}`}>
        <Link onClick={handleCardClick} href={`/video/${item.slug}`} className="">
          <Image
            src={(item.thumbnail?.templateUrl || '').replace(
              '{formatInstructions}',
              't_ratio16_9-size40/f_png'
            )}
            width="832"
            height="468"
            alt="Story Photo"
            className="w-full"
            placeholder="blur"
            blurDataURL={blurDataURL}
            loading="lazy"
          />
        </Link>
        <div className="md:absolute md:bottom-0 md:right-0 md:left-0 md:p-5 md:bg-black/70 md:text-white">
          <Link onClick={handleCardClick} href={`/video/${item.slug}`} className="">
            <h2 className="text-2xl font-bold">{item.headline || item.title}</h2>
          </Link>
          <div
            className="text-justify line-clamp-3 text-sm"
            dangerouslySetInnerHTML={{ __html: formatMarkdownContent(item.summary) }}
          />
          <Link href={`/video/${item.slug}`} className="block font-bold py-3 underline">
            Read Story
          </Link>
        </div>
      </div>
    );
  }

  return (
    <div className={className}>
      <p className="hidden" suppressHydrationWarning>
        {formatLocalizedDate(item.contentDate)} {formatLocalizedTime(item.contentDate)}
      </p>
      <Link onClick={handleCardClick} href={`/video/${item.slug}`} className="">
        {item.thumbnail ? (
          <Image
            src={(item.thumbnail?.templateUrl || '').replace(
              '{formatInstructions}',
              't_ratio16_9-size20/f_png'
            )}
            width="832"
            height="468"
            alt="Story Photo"
            className="mb-2 w-full"
            placeholder="blur"
            blurDataURL={blurDataURL}
            loading="lazy"
          />
        ) : (
          <MissingThumbnail />
        )}
        <h2 className="text-xl font-bold">{item.headline || item.title}</h2>
        <div className="text-sm mt-3">{item.fields.description}</div>
      </Link>
    </div>
  );
};

export default VideoCard;
