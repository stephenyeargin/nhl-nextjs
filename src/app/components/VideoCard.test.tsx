import React from 'react';
import { render, screen } from '@testing-library/react';
import VideoCard from './VideoCard';

jest.mock('next/image', () => {
  // eslint-disable-next-line @next/next/no-img-element, react/display-name
  const Img = ({ alt }: any) => <img alt={alt} />;
  (Img as any).displayName = 'NextImageMock';

  return Img;
});

// Make date/time formatting deterministic across environments
jest.mock('../utils/formatters', () => ({
  ...jest.requireActual('../utils/formatters'),
  // Match formatter defaults: date uses 'l' (M/D/YYYY), time uses 'LT z' (h:mm A z)
  // With TZ set to ET, 12:00 PM UTC becomes 8:00 AM ET on 3/10/2024
  formatLocalizedDate: () => '3/10/2024',
  formatLocalizedTime: () => '8:00 AM EDT',
}));

const baseItem = (overrides: any = {}) => ({
  slug: 'great-goal',
  headline: 'Great Goal',
  title: 'Great Goal',
  fields: { description: 'desc' },
  contentDate: '2024-03-10T12:00:00Z',
  summary: 'Some **markdown**',
  thumbnail: undefined,
  ...overrides,
});

describe('VideoCard', () => {
  it('renders default (no size) variant with fallback thumbnail', () => {
    const item = baseItem();
    const { asFragment } = render(<VideoCard item={item} />);
    expect(screen.getByText(/great goal/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders small variant with provided thumbnail url', () => {
    const item = baseItem({ thumbnail: { thumbnailUrl: 'thumb.jpg', templateUrl: '' } });
    const { asFragment } = render(<VideoCard item={item} size="small" />);
    expect(screen.getByText(/great goal/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders medium variant missing thumbnail', () => {
    const item = baseItem();
    const { asFragment } = render(<VideoCard item={item} size="medium" />);
    expect(screen.getByText(/great goal/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders large variant with template image', () => {
    const item = baseItem({ thumbnail: { templateUrl: 'http://x/{formatInstructions}/foo.png' } });
    const { asFragment } = render(<VideoCard item={item} size="large" />);
    expect(screen.getByText(/great goal/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
