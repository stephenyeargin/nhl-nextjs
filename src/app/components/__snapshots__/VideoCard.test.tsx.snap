// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoCard renders default (no size) variant with fallback thumbnail 1`] = `
<DocumentFragment>
  <div>
    <p
      class="hidden"
    >
      3/10/2024 8:00 AM EDT
    </p>
    <a
      class=""
      href="/video/great-goal"
    >
      <div
        class="w-full bg-gray-200 flex items-center justify-center aspect-square mb-2 aspect-video"
      >
        <img
          alt="Story Photo"
        />
      </div>
      <h2
        class="text-xl font-bold"
      >
        Great Goal
      </h2>
      <div
        class="text-sm mt-3"
      >
        desc
      </div>
    </a>
  </div>
</DocumentFragment>
`;

exports[`VideoCard renders large variant with template image 1`] = `
<DocumentFragment>
  <div
    class="md:relative undefined"
  >
    <a
      class=""
      href="/video/great-goal"
    >
      <img
        alt="Story Photo"
      />
    </a>
    <div
      class="md:absolute md:bottom-0 md:right-0 md:left-0 md:p-5 md:bg-black/70 md:text-white"
    >
      <a
        class=""
        href="/video/great-goal"
      >
        <h2
          class="text-2xl font-bold"
        >
          Great Goal
        </h2>
      </a>
      <div
        class="text-justify line-clamp-3 text-sm"
      >
        <p
          class="mb-4"
        >
          Some 
          <strong>
            markdown
          </strong>
        </p>
        

      </div>
      <a
        class="block font-bold py-3 underline"
        href="/video/great-goal"
      >
        Read Story
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`VideoCard renders medium variant missing thumbnail 1`] = `
<DocumentFragment>
  <div>
    <p
      class="hidden"
    >
      3/10/2024 8:00 AM EDT
    </p>
    <a
      class=""
      href="/video/great-goal"
    >
      <div
        class="w-full bg-gray-200 flex items-center justify-center aspect-square mb-2 aspect-square"
      >
        <img
          alt="Story Photo"
        />
      </div>
    </a>
    <a
      class=""
      href="/video/great-goal"
    >
      <h2
        class="text-xl font-bold"
      >
        Great Goal
      </h2>
    </a>
    <div
      class="text-justify line-clamp-3 text-sm"
    >
      <p
        class="mb-4"
      >
        Some 
        <strong>
          markdown
        </strong>
      </p>
      

    </div>
    <a
      class="block font-bold py-3 underline"
      href="/video/great-goal"
    >
      Watch
    </a>
  </div>
</DocumentFragment>
`;

exports[`VideoCard renders small variant with provided thumbnail url 1`] = `
<DocumentFragment>
  <div>
    <a
      class="grid grid-cols-3 gap-4 items-center"
      href="/video/great-goal"
    >
      <div
        class="col-span-1"
      >
        <img
          alt="Story Photo"
        />
      </div>
      <div
        class="col-span-2"
      >
        <h2
          class="text-base font-bold"
        >
          Great Goal
        </h2>
      </div>
    </a>
  </div>
</DocumentFragment>
`;
